### PR TITLE
Fix manifold copying on lines when converting hexes to tets.

### LINKED ISSUE #13040: weird manifold id settings when converting to simplices
I saw this when looking closely at #13035:

bad

This should be a cylinder, but it isn't (the top left and bottom right are wrong). It looks like we are not setting two line manifold ids correctly. Here is the code:
```cpp
#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_out.h>
#include <deal.II/grid/tria.h>

#include <fstream>

template <int dim> void check()
{
  using namespace dealii;
  Triangulation<dim> triangulation;

  {
    Triangulation<dim> x;
    GridGenerator::subdivided_cylinder(x, 4, 1, 2);

    GridGenerator::convert_hypercube_to_simplex_mesh(x, triangulation);
    for (const auto i : x.get_manifold_ids())
      if (i != numbers::flat_manifold_id)
        triangulation.set_manifold(i, x.get_manifold(i));
    triangulation.refine_global(3);
  }

  {
    GridOut go;
    std::ofstream out("out.vtk");
    go.write_vtk(triangulation, out);
  }
}

int main() { check<3>(); }
```

### PR DESCRIPTION
**Pull request by drwells**
Fixes #13040.

When we convert hexes to tets, we create three types of lines:
1. lines inside the hex
2. lines on the faces of the hex
3. lines equal to the hex's lines

In the last case we need to copy the manifold and boundary data from the preexisting lines directly, since they may not be equal to the ids assigned to the adjacent face.

Here's a picture of the bug in action. The top cell gets assigned a flat manifold id along one edge from the flat face instead of the curved face, so refinement produces the wrong result:
visit0000

fixed version with this patch:
visit0001

### CODE REVIEW

File: source/grid/grid_generator.cc [hunk 1]
```diff
@@ -7979,28 +7942,55 @@ namespace GridGenerator
             const auto bid = cell.face(f)->boundary_id();
             const auto mid = cell.face(f)->manifold_id();
 
+            // TODO - this isn't right. A line can have its own manifold id in
+            // 3D independent of whatever the face manifold ID is. That's where
+            // the bug comes from!
```
Comment (bangerth) [inline source/grid/grid_generator.cc:7947]: It isn't clear to me what "this" in "this isn't right" refers to. Is it something above or below? Can you be more specific?

File: source/grid/grid_generator.cc [hunk 2]
```diff
@@ -7979,28 +7942,55 @@ namespace GridGenerator
...
                 for (const auto &edge_vertices :
-                     vertex_ids_for_boundary_edges_3d[f])
+                     vertex_ids_for_new_boundary_edges_3d[f])
                   add_cell(1, edge_vertices, bid, mid);
               }
             else
               Assert(false, ExcNotImplemented());
           }
+
+        // set manifold ids of faces that were already present in the
```
Comment (bangerth) [inline source/grid/grid_generator.cc:7973]: do you mean 'edges' instead of 'faces' here?

File: source/grid/grid_generator.cc [hunk 3]
```diff
@@ -7979,28 +7942,55 @@ namespace GridGenerator
...
+        // triangulation.
+        if (dim == 3)
+          {
+            for (const auto e : cell.line_indices())
+              {
+                auto edge = cell.line(e);
+                // Since we are just copying information and not adding anything
+                // new, rather than translate back-and-forth from local and
+                // global indices just add to subcell data directly.
```
Comment (bangerth) [inline source/grid/grid_generator.cc:7982]: ...instead of using `add_cell()`?

**Review by bangerth — COMMENTED** [review]
Just a couple of small issues.

File: source/grid/grid_generator.cc (hunk 1 above)
Comment (drwells) [reply source/grid/grid_generator.cc:7947]: that was a comment to myself that I meant to delete - the bug is the subject of the PR.

Comment (bangerth) [reply source/grid/grid_generator.cc:7947]: OK. Let me know once you push an updated version!

**Comment by bangerth** [comment]
Maybe also a changelog entry?

**Comment by drwells** [comment]
Fixed!

**Review by bangerth — APPROVED** [review]

**Pull request merged**